src/logger.ts

```diff
@@ -0,0 +1,432 @@
+		uuid: string,
+		deviceApiKey: string,
+	) {
+		super();
+
+		this.opts = url.parse(`${apiEndpoint}/device/v2/${uuid}/log-stream`) as any;
+		this.opts.method = 'POST';
+		this.opts.headers = {
+			Authorization: `Bearer ${deviceApiKey}`,
+			'Content-Type': 'application/x-ndjson',
+			'Content-Encoding': 'gzip',
+		};
+
+		// This stream serves serves as a message buffer during reconnections
+		// while we unpipe the old, malfunctioning connection and then repipe a
+		// new one.
+		this.stream = new stream.PassThrough({
+			allowHalfOpen: true,
+
+			// We halve the high watermark because a passthrough stream has two
+			// buffers, one for the writable and one for the readable side. The
+			// write() call only returns false when both buffers are full.
+			highWaterMark: MAX_PENDING_BYTES / 2,
+		});
+
+		this.stream.on('drain', () => {
+			this.writable = true;
+			this.flush();
+			if (this.dropCount > 0) {
+				this.write({
+					message: `Warning: Suppressed ${this.dropCount} message(s) due to high load`,
+					timestamp: Date.now(),
+					isSystem: true,
+					isStdErr: true,
+				});
+				this.dropCount = 0;
+			}
+		});
+	}
+
+	public log(message: LogMessage) {
+		if (this.offlineMode || !this.publishEnabled) {
+			return;
+		}
+
+		if (!_.isObject(message)) {
+			return;
+		}
+
+		message = _.assign({
+			timestamp: Date.now(),
+			message: '',
+		}, message);
+
+		if (!(message.isSystem || message.serviceId != null)) {
+			return;
+		}
+
+		message.message = _.truncate(message.message, {
+			length: MAX_LOG_LENGTH,
+			omission: '[...]',
+		});
+
+		this.write(message);
+	}
+
+	private setup = _.throttle(() => {
+		this.req = https.request(this.opts);
+
+		// Since we haven't sent the request body yet, and never will,the
+		// only reason for the server to prematurely respond is to
+		// communicate an error. So teardown the connection immediately
+		this.req.on('response', (res) => {
+			console.log('LogBackend: server responded with status code:', res.statusCode);
+			this.teardown();
+		});
+
+		this.req.on('timeout', () => this.teardown());
+		this.req.on('close', () => this.teardown());
+		this.req.on('error', (err) => {
+			console.log('LogBackend: unexpected error:', err);
+			this.teardown();
+		});
+
+		// Immediately flush the headers. This gives a chance to the server to
+		// respond with potential errors such as 401 authentication error
+		this.req.flushHeaders();
+
+
+		// We want a very low writable high watermark to prevent having many
+		// chunks stored in the writable queue of @_gzip and have them in
+		// @_stream instead. This is desirable because once @_gzip.flush() is
+		// called it will do all pending writes with that flush flag. This is
+		// not what we want though. If there are 100 items in the queue we want
+		// to write all of them with Z_NO_FLUSH and only afterwards do a
+		// Z_SYNC_FLUSH to maximize compression
+		this.gzip = zlib.createGzip({ writableHighWaterMark: 1024 } as any);
```

Comment:
Is this `any` definitely necessary?

Reply:
Yeah, the typings for zlib don't include `writableHighWaterMark`.

Reply:
Have you tried using the node 6 typings?  It's possible they'll include it (and we should be using those instead to not accidentally try to use node 10 things).  Aside from that you should be able to extend the interface with
```typescript
declare module 'zlib' {
	export interface ZlibOptions {
		writableHighWaterMark: number;
	}
}